src/app/core/http/interceptors/refresh-jwt.interceptor.ts

```diff
@@ -0,0 +1,70 @@
+import { HttpErrorResponse, HttpEvent, HttpHandlerFn, HttpInterceptorFn, HttpRequest } from '@angular/common/http'
+import { catchError, Observable, throwError } from 'rxjs'
+import { inject } from '@angular/core'
+import { AuthService } from '@core/auth/services/auth.service'
+import { TokensStorageService } from '@core/auth/services/tokens-storage.service'
+import { Store } from '@ngxs/store'
+import { ToastrService } from 'ngx-toastr'
+import { LogOut, SetTokens } from '@core/auth/state/auth.actions'
+import { switchMap } from 'rxjs/operators'
+
+export const refreshJwtInterceptor: HttpInterceptorFn = (
+  req: HttpRequest<unknown>,
+  next: HttpHandlerFn
+): Observable<HttpEvent<unknown>> => {
+  const apiService = inject(AuthService)
+  const tokenStorage = inject(TokensStorageService)
+  const store = inject(Store)
+  const toastr = inject(ToastrService)
+
+  return next(req).pipe(
+    catchError((error: HttpErrorResponse) => {
+      if (error.status === 401 && !req.url.includes('token/new')) {
+        return handle401Error(req, next)
+      }
+
+      if (error.status === 403) {
+        return handle403Error(req, next)
+      }
+
+      return throwError(() => error)
+    })
+  )
+
+  function handle401Error(
+    request: HttpRequest<unknown>,
+    nextFn: HttpHandlerFn
+  ): Observable<HttpEvent<unknown>> {
+    const token = tokenStorage.get()?.refreshToken
+
+    if (!token) {
+      store.dispatch(new LogOut())
+
+      return nextFn(req)
+    }
+
+    return apiService.refreshToken(token).pipe(
+      switchMap(tokens => {
+        store.dispatch(new SetTokens(tokens))
+
+        return nextFn(req)
+      }),
+      catchError((err: HttpErrorResponse) => {
+        return throwError(() => err)
+      })
+    )
+  }
+
+  function handle403Error(
+    request: HttpRequest<unknown>,
+    nextFn: HttpHandlerFn
+  ): Observable<HttpEvent<unknown>> {
+    store.dispatch(new LogOut())
```

Comment:
Why would you log out a user, who is legitimately logged in, but just accessed a resource they don't permission to? They still have permissions to view a lot of other things in your app

Reply:
I didn't think about the case with view possibility. I thought, if a user doesn't have permission to make a web API call it means he should be logged out automatically.